data/transactions/logic/TEAL_opcodes.md

```diff
@@ -464,6 +464,7 @@ FirstValidTime causes the program to fail. The field is reserved for future use.
 | 7 | LatestTimestamp | uint64 | Last confirmed block UNIX timestamp. Fails if negative. LogicSigVersion >= 2. |
 | 8 | CurrentApplicationID | uint64 | ID of current application executing. Fails if no such application is executing. LogicSigVersion >= 2. |
 | 9 | CreatorAddress | []byte | Address of the creator of the current application. Fails if no such application is executing. LogicSigVersion >= 3. |
+| 10 | CurrentApplicationAddress | []byte | Address that the current application controls. Fails if no such application is executing. LogicSigVersion >= 5. |
```

Comment:
Adding a controlled address is cool. Is there any application benefit for supporting multiple application controlled accounts for a single application ?

Reply:
There is at least one - suppose you wanted to escrow algos from many accounts, and always be wiling to give them back, with the rewards they've earned in the meantime.  Since rewards are impossible to calculate in contract, you'd need to segregate your deposits, rather than keep them in one pot.

I originally planned on having a way to create any number of app accounts, for exactly this reason (and some more obscure ideas). But I think that the ability to rekey to the app addr is powerful enough.  To use such a contract, a depositor would make a new account, opt it into the contract, and then, in an atomic group:
1. make a deposit
2. call the contract to bind the new address to the depositor (by writing the deposit address in address's app local state)
3. rekey the new account to the app account

Now, when depositing or withdrawing, the caller tells the contract where his deposit account is.  So the contract can use these extra accounts that were ceded over to it because of rekeying. Which also prevents the caller from messing with them ever again.

(If you really like cleanliness, you'd have an operation to ask the contract to close out the account if you wanted to withdraw it all.)